File: 4.Stack/java/NextGreaterElement.java
/**
 * Next Greater Element using Stack in Java
 * 
 * This program demonstrates how to find the next greater element for each element
 * in an array using a stack. For each element, the next greater element is the first
 * greater element to its right. If there is no greater element, the result is -1.
 * 
 * Time Complexity: O(n) where n is the size of the array
 * Space Complexity: O(n) for the stack and result array
 */

package java;

import java.util.Arrays;
import java.util.Stack;

public class NextGreaterElement {
    
    /**
     * Find the next greater element for each element in the array
     * (Processing from right to left)
     * 
     * @param arr The input array
     * @return An array containing the next greater element for each element
     */
    public static int[] nextGreaterElement(int[] arr) {
        int n = arr.length;
        int[] result = new int[n];
        Arrays.fill(result, -1); // Initialize with -1 (no greater element)
        
        Stack<Integer> stack = new Stack<>(); // Stack to store elements
        
        // Process all elements from right to left
        for (int i = n - 1; i >= 0; i--) {
            // Remove elements from stack that are smaller than or equal to current element
            while (!stack.isEmpty() && stack.peek() <= arr[i]) {
                stack.pop();
            }
            
            // If stack is not empty, the top element is the next greater element
            if (!stack.isEmpty()) {
                result[i] = stack.peek();
            }
            
            // Push current element to stack
            stack.push(arr[i]);
        }
        
        return result;
    }
    
    /**
     * Find the next greater element for each element in the array
     * (Processing from left to right)
     * 
     * @param arr The input array
     * @return An array containing the next greater element for each element
     */
    public static int[] nextGreaterElementLeftToRight(int[] arr) {
        int n = arr.length;
        int[] result = new int[n];
        Arrays.fill(result, -1); // Initialize with -1 (no greater element)
        
        Stack<Integer> stack = new Stack<>(); // Stack to store indices
        
        // Process all elements from left to right
        for (int i = 0; i < n; i++) {
            // Remove indices from stack whose elements are smaller than current element
            while (!stack.isEmpty() && arr[stack.peek()] < arr[i]) {
                // For each popped index, the current element is the next greater element
                result[stack.pop()] = arr[i];
            }
            
            // Push current index to stack
            stack.push(i);
        }
        
        return result;
    }
    
    /**
     * Print an array with its next greater elements
     * 
     * @param arr The original array
     * @param nge The array of next greater elements
     * @param title The title for the output
     */
    public static void printNextGreaterElements(int[] arr, int[] nge, String title) {
        System.out.println(title);
        System.out.println("Element\tNGE");
        for (int i = 0; i < arr.length; i++) {
            System.out.println(arr[i] + "\t" + nge[i]);
        }
        System.out.println();
    }
    
    /**
     * Main method to demonstrate the next greater element algorithms
     */
    public static void main(String[] args) {
        System.out.println("Next Greater Element using Stack in Java");
        System.out.println("---------------------------------------");
        
        // Example 1: Basic array
        int[] arr1 = {4, 5, 2, 25, 7, 8};
        System.out.println("\nExample 1: Basic array");
        System.out.print("Array: ");
        for (int num : arr1) {
            System.out.print(num + " ");
        }
        System.out.println();
        
        int[] result1 = nextGreaterElement(arr1);
        printNextGreaterElements(arr1, result1, "\nRight to Left Approach:");
        
        int[] result1Alt = nextGreaterElementLeftToRight(arr1);
        printNextGreaterElements(arr1, result1Alt, "Left to Right Approach:");
        
        // Example 2: Array with duplicates
        int[] arr2 = {11, 13, 21, 3, 4, 3, 21, 34};
        System.out.println("\nExample 2: Array with duplicates");
        System.out.print("Array: ");
        for (int num : arr2) {
            System.out.print(num + " ");
        }
        System.out.println();
        
        int[] result2 = nextGreaterElement(arr2);
        printNextGreaterElements(arr2, result2, "\nRight to Left Approach:");
        
        int[] result2Alt = nextGreaterElementLeftToRight(arr2);
        printNextGreaterElements(arr2, result2Alt, "Left to Right Approach:");
        
        // Example 3: Array in descending order (worst case)
        int[] arr3 = {5, 4, 3, 2, 1};
        System.out.println("\nExample 3: Array in descending order (worst case)");
        System.out.print("Array: ");
        for (int num : arr3) {
            System.out.print(num + " ");
        }
        System.out.println();
        
        int[] result3 = nextGreaterElement(arr3);
        printNextGreaterElements(arr3, result3, "\nRight to Left Approach:");
        
        int[] result3Alt = nextGreaterElementLeftToRight(arr3);
        printNextGreaterElements(arr3, result3Alt, "Left to Right Approach:");
        
        // Example 4: Array in ascending order
        int[] arr4 = {1, 2, 3, 4, 5};
        System.out.println("\nExample 4: Array in ascending order");
        System.out.print("Array: ");
        for (int num : arr4) {
            System.out.print(num + " ");
        }
        System.out.println();
        
        int[] result4 = nextGreaterElement(arr4);
        printNextGreaterElements(arr4, result4, "\nRight to Left Approach:");
        
        int[] result4Alt = nextGreaterElementLeftToRight(arr4);
        printNextGreaterElements(arr4, result4Alt, "Left to Right Approach:");
        
        // Example 5: Array with a single element
        int[] arr5 = {10};
        System.out.println("\nExample 5: Array with a single element");
        System.out.print("Array: ");
        for (int num : arr5) {
            System.out.print(num + " ");
        }
        System.out.println();
        
        int[] result5 = nextGreaterElement(arr5);
        printNextGreaterElements(arr5, result5, "\nRight to Left Approach:");
        
        int[] result5Alt = nextGreaterElementLeftToRight(arr5);
        printNextGreaterElements(arr5, result5Alt, "Left to Right Approach:");
        
        // Performance comparison note
        System.out.println("\nPerformance Note:");
        System.out.println("Both approaches have the same time complexity O(n) and space complexity O(n).");
        System.out.println("The Left to Right approach might be more intuitive for some problems.");
        System.out.println("The Right to Left approach can be more concise for certain implementations.");
    }
}

File: 4.Stack/Kotlin/NextGreaterElement.kt
/**
 * Next Greater Element Problem in Kotlin
 * 
 * This program demonstrates how to find the next greater element for each element in an array.
 * For each element, the next greater element is the first greater element to its right.
 * If there is no greater element, the result is -1.
 * 
 * Two approaches are implemented:
 * 1. Using a stack and processing the array from right to left
 * 2. Using a stack and processing the array from left to right
 * 
 * Time Complexity: O(n) where n is the size of the array
 * Space Complexity: O(n) for the stack and result array
 */

package Kotlin

import java.util.Stack

/**
 * Find the next greater element for each element in the array
 * Processing the array from right to left
 * 
 * @param arr The input array
 * @return An array containing the next greater element for each element
 */
fun nextGreaterElement(arr: IntArray): IntArray {
    val n = arr.size
    val result = IntArray(n) { -1 } // Initialize with -1 (no greater element)
    val stack = Stack<Int>()
    
    // Process all elements from right to left
    for (i in n - 1 downTo 0) {
        // Remove elements from stack that are smaller than or equal to current element
        while (stack.isNotEmpty() && stack.peek() <= arr[i]) {
            stack.pop()
        }
        
        // If stack is not empty, the top element is the next greater element
        if (stack.isNotEmpty()) {
            result[i] = stack.peek()
        }
        
        // Push current element to stack
        stack.push(arr[i])
    }
    
    return result
}

/**
 * Find the next greater element for each element in the array
 * Processing the array from left to right
 * 
 * @param arr The input array
 * @return An array containing the next greater element for each element
 */
fun nextGreaterElementLeftToRight(arr: IntArray): IntArray {
    val n = arr.size
    val result = IntArray(n) { -1 } // Initialize with -1 (no greater element)
    val stack = Stack<Int>() // Stack to store indices, not values
    
    // Process all elements from left to right
    for (i in 0 until n) {
        // While stack is not empty and current element is greater than element at stack top
        while (stack.isNotEmpty() && arr[i] > arr[stack.peek()]) {
            // For the element at stack top, the next greater element is the current element
            val idx = stack.pop()
            result[idx] = arr[i]
        }
        
        // Push current index to stack
        stack.push(i)
    }
    
    // Elements remaining in stack have no next greater element
    // (result array is already initialized with -1)
    
    return result
}

/**
 * Print the array and its next greater elements
 * 
 * @param arr The original array
 * @param result The array containing next greater elements
 * @param approach The name of the approach used
 */
fun printNextGreaterElements(arr: IntArray, result: IntArray, approach: String) {
    println("$approach:")
    println("Element\tNext Greater")
    for (i in arr.indices) {
        println("${arr[i]}\t\t${result[i]}")
    }
    println()
}

/**
 * Main function to demonstrate the next greater element problem
 */
fun main() {
    println("Next Greater Element Problem in Kotlin")
    println("-------------------------------------")
    
    // Example 1: Basic array
    val arr1 = intArrayOf(4, 5, 2, 10, 8)
    println("\nExample 1: ${arr1.joinToString()}")
    
    val result1RightToLeft = nextGreaterElement(arr1)
    printNextGreaterElements(arr1, result1RightToLeft, "Right to Left Approach")
    
    val result1LeftToRight = nextGreaterElementLeftToRight(arr1)
    printNextGreaterElements(arr1, result1LeftToRight, "Left to Right Approach")
    
    // Example 2: Array with duplicates
    val arr2 = intArrayOf(4, 5, 5, 2, 25, 25)
    println("\nExample 2: ${arr2.joinToString()}")
    
    val result2RightToLeft = nextGreaterElement(arr2)
    printNextGreaterElements(arr2, result2RightToLeft, "Right to Left Approach")
    
    val result2LeftToRight = nextGreaterElementLeftToRight(arr2)
    printNextGreaterElements(arr2, result2LeftToRight, "Left to Right Approach")
    
    // Example 3: Array in descending order (no next greater element)
    val arr3 = intArrayOf(10, 8, 6, 4, 2)
    println("\nExample 3: ${arr3.joinToString()}")
    
    val result3RightToLeft = nextGreaterElement(arr3)
    printNextGreaterElements(arr3, result3RightToLeft, "Right to Left Approach")
    
    val result3LeftToRight = nextGreaterElementLeftToRight(arr3)
    printNextGreaterElements(arr3, result3LeftToRight, "Left to Right Approach")
    
    // Example 4: Array in ascending order
    val arr4 = intArrayOf(1, 3, 5, 7, 9)
    println("\nExample 4: ${arr4.joinToString()}")
    
    val result4RightToLeft = nextGreaterElement(arr4)
    printNextGreaterElements(arr4, result4RightToLeft, "Right to Left Approach")
    
    val result4LeftToRight = nextGreaterElementLeftToRight(arr4)
    printNextGreaterElements(arr4, result4LeftToRight, "Left to Right Approach")
    
    // Example 5: Array with a single element
    val arr5 = intArrayOf(5)
    println("\nExample 5: ${arr5.joinToString()}")
    
    val result5RightToLeft = nextGreaterElement(arr5)
    printNextGreaterElements(arr5, result5RightToLeft, "Right to Left Approach")
    
    val result5LeftToRight = nextGreaterElementLeftToRight(arr5)
    printNextGreaterElements(arr5, result5LeftToRight, "Left to Right Approach")
    
    // Performance comparison
    println("\nPerformance Comparison:")
    println("Both approaches have O(n) time complexity and O(n) space complexity.")
    println("The Left to Right approach is often more intuitive as it processes elements in order.")
    println("The Right to Left approach can be more efficient for certain patterns.")
    
    // Functional approach using Kotlin's features
    println("\nFunctional Approach:")
    val arr6 = intArrayOf(4, 5, 2, 10, 8)
    println("Original array: ${arr6.joinToString()}")
    
    // Using extension function for better readability
    fun IntArray.findNextGreaterElements(): List<Pair<Int, Int>> {
        val result = nextGreaterElementLeftToRight(this)
        return this.zip(result.toList())
    }
    
    val pairs = arr6.findNextGreaterElements()
    println("Element -> Next Greater:")
    pairs.forEach { (element, nextGreater) ->
        println("$element -> $nextGreater")
    }
    
    // Using higher-order functions
    println("\nElements with no next greater element:")
    pairs.filter { (_, nextGreater) -> nextGreater == -1 }
         .map { (element, _) -> element }
         .also { println(it) }
    
    println("\nElements with next greater element > 2x their value:")
    pairs.filter { (element, nextGreater) -> nextGreater > element * 2 }
         .map { (element, nextGreater) -> "$element -> $nextGreater" }
         .also { println(it) }
}